### ISSUE TITLE
Code accumulating memory after migration to v1.0.0

### ISSUE THREAD

**Issue by tomchor** [JuliaLang/julia#29014]
I have [this code](https://github.com/tomchor/3dht) that I wrote in Julia v0.6. It worked perfectly with that version. I recently migrated it to v1.0.0. The results are the same, so in a way it works, but I notice when running the main loop that the RAM keeps increasing and increasing in use, which didn't use to happen with v0.6. Even after the loop is done, the RAM does not go back to what it was.

I think have isolated the behavior to [this function](https://github.com/tomchor/3dht/blob/master/aux_schemes.jl#L3L36), but I see no reason for this to happen. When I issue `varinfo()` I see nothing unusual: all the variables I have are pretty small in memory (the largest variable I have is 2.5 MB, while the RAM builds up until all 8GB are full).

Since I see nothing in `varinfo()` and since this didn't happen in v0.6, I'm submitting this as a bug report. If anyone wants to try my code, here's all you have to do to reproduce the behavior:

```
include("main.jl")
run_sim(Uh, NL);
```

PS.: I'm sorry but so far I could not make a simple MWE for this. I haven't been able to isolate the source.

Here's some relevant info:

```
Julia Version 1.0.0
Commit 5d4eaca0c9 (2018-08-08 20:58 UTC)
Platform Info:
  OS: Linux (x86_64-pc-linux-gnu)
  CPU: Intel(R) Core(TM) i5-7200U CPU @ 2.50GHz
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-6.0.0 (ORCJIT, skylake)

```

**Comment by KristofferC** [comment]
Do you actually run out of memory if you run it long enough?

What if you explicitly do `GC.gc(); GC.gc(); GC.gc()`, does it reclaim any memory?

**Comment by tomchor** [comment]
Yes, if I run long enough everything crashes and I have to manually kill Julia.

I just tried putting `GC.gc()` as the last line of the loop but the results are the same. I couldn't notice any difference.

**Comment by tomchor** [comment]
@KristofferC By the way, in case it is not clear, even after my script stops running I can't get rid of the extra memory. The only way I can do it (from the ways I've tried so far) is by closing Julia and opening it up again.

Can someone at least confirm if this is indeed a bug?

**Comment by tshort** [comment]
Might be related to the following:

https://github.com/JuliaPy/PyCall.jl/issues/542

It's probably unrelated to the bug, but for performance reasons, you may want to stop using non-constant global variables. To make it easier for others to try, you could make it into a package.

**Comment by pablosanjose** [comment]
It is possibly not the real issue, but note that in your code each time you do `variable[n,:,:,:]` you are allocating an array, unless you are in a broadcasted assignment `var1[n,:,:,:] .= func.(var2[m,:,:,:])` or similar. First thing I would try is to use `view` for the slices or do `.=` instead of `=`. Although it's true that all those slices should not be leaking memory in any case...

**Comment by tomchor** [comment]
@pablosanjose Thanks for the tip. I followed your suggestion and I think there was an improve in efficiency (but I'm not 100% sure since I didn't run any rigorous tests) but there was no difference in the memory leak.

@tshort It appears that the issue is indeed PyCall. I have replaced calls for `np.roll` with `circshift` (which I just found out exists) and the issue appears to have significantly decreased. I still need to use PyCall to write results inside the loop, so the issue hasn't yet disappeared for me, but it's good to at least know where it's coming from now.
